cartographer/mapping/map_builder.cc

```diff
@@ -218,6 +218,14 @@ void MapBuilder::LoadMap(io::ProtoStreamReader* const reader) {
     if (!reader->ReadProto(&proto)) {
       break;
     }
+    if (proto.has_node()) {
+      const auto& pose_graph_node =
+          pose_graph.trajectory(proto.node().node_id().trajectory_id())
+              .node(proto.node().node_id().node_index());
+      const transform::Rigid3d pose =
+          transform::ToRigid3(pose_graph_node.pose());
+      sparse_pose_graph_->AddNodeFromProto(pose, proto.node());
```

Comment:
To get this merged, let's pass the `map_trajectory_id` similar to the handling of submaps. Map loading right now creates a single trajectory to contain the frozen data. This is unproblematic since no optimization variables are added for the frozen data.

To handle multiple trajectories with trimmed data, we have to change the serialization first. Since we now also want to support trimming in the middle, we have to change the serialization to maintain information about gaps, e.g. having explicit node/submap indices instead of implicit from the repeated field index.

Reply:
I understand that we can't use the full node/submap ID from the proto, because the node/submap index is not remapped to account for trimmed nodes/submaps. But I am confused about why can't we use just the trajectory ID from the proto? AFAIK these are not affected by trimming. I don't see an advantage to passing `map_trajectory_id` via the function call, same with `AddSubmapFromProto`.

And that `CHECK_EQ(node_id, proto_node_id)` which breaks loading trimmed trajectories can be disabled, as I suggested in #567, or softened just to compare the trajectory IDs.

Reply:
We can't just use the ID from the proto because we only create one frozen trajectory here which has the ID `map_trajectory_id`. If we merge this PR now, we might add nodes to trajectories which do not exist, i.e. this is a bug. If you pass the `map_trajectory_id`, we this PR adds support for loading maps including the node data to use it for loop closures.

Reply:
Yes, that's why imagined that this gets merged after #567. But sure, if you want to merge this first, I can pass `map_trajectory_id` for now.

Reply:
Yes, this is preferable so that the code base is in a good state at each commit if possible.